### ISSUE TITLE
Specific rebuild sequence throws 'Multiple widgets used the same GlobalKey' exception

### ISSUE THREAD

**Issue by krisgiesing** [flutter/flutter#43780]
The test at https://gist.github.com/krisgiesing/afd4c3bd82458f8eeba847fe2cc7b08d appears to be legal. However, running it throws the following exception:

```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown building _Stateful-[GlobalKey#bdeee 2](state:
_StatefulState#6790b):
Multiple widgets used the same GlobalKey [GlobalObjectKey String#77ca8].
The key [GlobalObjectKey String#77ca8] was used by multiple widgets. The parents of those widgets
were:
- _Stateful-[GlobalKey#e2abd 1](state: _StatefulState#5e00d)
- _Stateful-[GlobalKey#bdeee 2](state: _StatefulState#6790b)
A GlobalKey can only be specified on one widget at a time in the widget tree.

Widget creation tracking is currently disabled. Enabling it enables improved error messages. It can
be enabled by passing `--track-widget-creation` to `flutter run` or `flutter test`.

When the exception was thrown, this was the stack:
#0      GlobalKey._debugReserveFor.<anonymous closure> (package:flutter/src/widgets/framework.dart:189:11)
#1      GlobalKey._debugReserveFor (package:flutter/src/widgets/framework.dart:212:6)
#2      Element.updateChild.<anonymous closure> (package:flutter/src/widgets/framework.dart:2905:13)
#3      Element.updateChild (package:flutter/src/widgets/framework.dart:2909:6)
#4      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4016:16)
#5      Element.rebuild (package:flutter/src/widgets/framework.dart:3793:5)
#6      StatefulElement.update (package:flutter/src/widgets/framework.dart:4181:5)
#7      Element.updateChild (package:flutter/src/widgets/framework.dart:2926:15)
#8      SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5210:14)
#9      Element.updateChild (package:flutter/src/widgets/framework.dart:2926:15)
#10     RenderObjectElement.updateChildren (package:flutter/src/widgets/framework.dart:4919:32)
#11     MultiChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:5319:17)
#12     Element.updateChild (package:flutter/src/widgets/framework.dart:2926:15)
#13     _LayoutBuilderElement._layout.<anonymous closure> (package:flutter/src/widgets/layout_builder.dart:115:18)
#14     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2355:19)
#15     _LayoutBuilderElement._layout (package:flutter/src/widgets/layout_builder.dart:95:11)
#16     RenderObject.invokeLayoutCallback.<anonymous closure> (package:flutter/src/rendering/object.dart:1815:58)
#17     PipelineOwner._enableMutationsToDirtySubtrees (package:flutter/src/rendering/object.dart:875:15)
#18     RenderObject.invokeLayoutCallback (package:flutter/src/rendering/object.dart:1815:13)
#19     RenderConstrainedLayoutBuilder.layoutAndBuildChild (package:flutter/src/widgets/layout_builder.dart:173:5)
#20     _RenderLayoutBuilder.performLayout (package:flutter/src/widgets/layout_builder.dart:240:5)
#21     RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1584:7)
#22     PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:844:18)
#23     AutomatedTestWidgetsFlutterBinding.drawFrame (package:flutter_test/src/binding.dart:974:23)
#24     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:279:5)
#25     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1043:15)
#26     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:982:9)
#27     AutomatedTestWidgetsFlutterBinding.pump.<anonymous closure> (package:flutter_test/src/binding.dart:869:9)
#30     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:69:41)
#31     AutomatedTestWidgetsFlutterBinding.pump (package:flutter_test/src/binding.dart:856:27)
#32     WidgetTester.pumpWidget.<anonymous closure> (package:flutter_test/src/widget_tester.dart:320:22)
#35     TestAsyncUtils.guard (package:flutter_test/src/test_async_utils.dart:69:41)
#36     WidgetTester.pumpWidget (package:flutter_test/src/widget_tester.dart:317:27)
#37     main.<anonymous closure>.<anonymous closure> (file:///Users/kgiesing/Code/turquoise/flutter/packages/flutter/test/material/key_test.dart:62:20)
#51     AutomatedTestWidgetsFlutterBinding.runTest.<anonymous closure> (package:flutter_test/src/binding.dart:1063:17)
#53     AutomatedTestWidgetsFlutterBinding.runTest.<anonymous closure> (package:flutter_test/src/binding.dart:1051:35)
(elided 33 frames from class _FakeAsync, package dart:async, package dart:async-patch, and package stack_trace)
```

**Comment by krisgiesing** [comment]
@Hixie fyi

**Comment by krisgiesing** [comment]
[Transcribed from chat]

As far as I can tell these are the necessary factors:
1) a state object keyed by index is marked as needing rebuild
2) its child is keyed by contents
3) the map of index to contents changes before the state gets to rebuild
4) the entire tree also needs to rebuild
5) the entire tree is built using LayoutBuilder
6) the entire tree rebuilds using the new map of index to contents, changing the relationship between the index-keyed parent and the content-keyed child

This creates a situation where the two buildScope passes are using different ideas of what parents different keys are reserved for. It would all be resolved by the time the second buildScope would be done, but the exception is thrown first.

**Comment by mtcliatt** [comment]
Does the Flutter team have insight into what might be causing this issue?

**Comment by krisgiesing** [comment]
It would be great to have a Flutter team member take enough of a look at this to indicate whether it's a framework bug or something else. @Hixie suggested that perhaps the check needs to be done in a different part of the build sequence to account for LayoutBuilder; if that's true then I'm happy to look into moving it, but it seems like the issue needs a little bit more analysis on the Flutter side to know if that's the right thing to do.

**Comment by goderbauer** [comment]
It's on my radar to look at this soon, I just haven't found the time for it yet.

**Comment by chunhtai** [comment]
Finished investigation.

I have a slightly simpler repo
``` dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

void main() {
  group('KeyTest', () {
    final List<String> listItems = <String>['Item 1', 'Item 2', 'Item 3', 'Item 4'];

    Widget listItemToWidget(int index, String listItem) {
      return _Stateful(
        child: Text(
          listItem,
          textDirection: TextDirection.ltr,
          key: GlobalObjectKey(listItem),
        ),
      );
    }

    Widget buildContents() {
      final List<Widget> children = <Widget>[];
      for (int i = 0; i < listItems.length; i++) {
        children.add(listItemToWidget(i, listItems[i]));
      }
      return Column(
        children: children,
      );
    }

    testWidgets('allows reordering inside the middle of the widget (reorder up)', (WidgetTester tester) async {
      await tester.pumpWidget(
        LayoutBuilder(
          builder: (_,__) => buildContents(),
        )
      );
      final _StatefulState state = tester.firstState(find.byType(_Stateful).at(1));
      state.rebuild();
      // Reorder the items
      final String temp = listItems[1];
      listItems[1] = listItems[2];
      listItems[2] = temp;
      await tester.pumpWidget(
        LayoutBuilder(
          builder: (_,__) => buildContents(),
        )
      );
    });
  });
}

class _Stateful extends StatefulWidget {
  const _Stateful({Key key, this.child}) : super(key: key);
  final Text child;
  @override
  State<StatefulWidget> createState() => _StatefulState();
}

class _StatefulState extends State<_Stateful> {
  void rebuild() => setState(() {});
  @override
  void didUpdateWidget(_Stateful oldWidget){
    super.didUpdateWidget(oldWidget);
    print('old widget ${oldWidget.child.data} to new widget ${widget.child.data}');
  }
  @override
  Widget build(BuildContext context) {
    return widget.child;
  }
  @override
  String toString({DiagnosticLevel minLevel = DiagnosticLevel.debug}) {
    return 'State ${widget.child.data}';
  }
}
```

This is a corner case that should only happens in layoutBuilder and a dirty children prior to rebuild
Consider this case
![image](https://user-images.githubusercontent.com/47866232/69761758-2e7eb100-111d-11ea-9502-a6699ab6920e.png)
After the state rebuild, reorder and pumpwidget. It become like this

![image](https://user-images.githubusercontent.com/47866232/69762068-00e63780-111e-11ea-9719-ef5ea58a837e.png)

During the widgetBinding drawframe, it will rebuild the dirty element.
The problem here is Layoutbuilder will push the subtree rebuild to rendering phase and continue to next dirty element which is child1.
The child 1 will then reserve globalkey 1 to child 1. At this pointer, the reserve map looks like this
```
{
   Key1 : ElementofChild1
}
```
Later on the Layoutbuilder rebuilds its subtree and it figures out the children are swapped. It will use didupdatewidget to update ElementofChild1 to become ElementofChild2*(ElementofChild1 and ElementofChild2* are the same reference but updated state) and reserve Key2 to ElementofChild2*
At this point, the reserve map looks like this
{
   Key1 : ElementofChild2*, // We are keeping the reference here. update the value of reference will 
                                             //also update here
   Key2 : ElementofChild2*
}

and the it will update  ElementofChild2 to ElementofChild1* and reserve Key1 for it, this is where the error is throw because Key1 is already reserved

**Comment by chunhtai** [comment]
@krisgiesing is this currently reproducible in an actual app, or this is only reproducible in the test.
Are you currently blocked by this issue?

**Comment by krisgiesing** [comment]
This blocks landing https://github.com/flutter/flutter/pull/41338/files, which fixes https://github.com/flutter/flutter/issues/41334. That issue is reproducible in our own app but is also easily reproduced by modifying the kitchen sink app. Note that the PR that fixes the issue includes a test that illustrates how a real app might run into this situation.

The impact to our app specifically is that the section of UI that is contained with ReorderableListView children is constantly rebuilding when it doesn't need to. The performance hit is noticeable because (in our case) the contents are nontrivial to compute.

**Comment by chunhtai** [comment]
@Hixie 
Do you know what was the goal for this check ?https://github.com/flutter/flutter/blob/f213f92587b3f6f3b079a9cf356b2c5fcf00d20b/packages/flutter/lib/src/widgets/framework.dart#L148
This check will already check if there are two or more elements shared the same global key
https://github.com/flutter/flutter/blob/f213f92587b3f6f3b079a9cf356b2c5fcf00d20b/packages/flutter/lib/src/widgets/framework.dart#L192

If we have second check, would the first one still be necessary?

**Comment by Hixie** [comment]
In general all of this logic is to catch cases where you have two widgets with the same GlobalKey in the tree at the same time.

There are many cases to consider:

- A widget with key X is in the tree. Later, an unrelated widget is marked dirty, rebuilds, and adds a new widget with key X (not the same type as the original X). Even though only one widget was added that frame, we must detect that there are two Xs in the tree (without a global tree walk which is prohibitively expensive).

- A widget with key Y is in the tree. An unrelated widget is marked dirty, and it adds a child with key Y, having the same type as the original widget with key Y, causing the original Y's element to be reparented. The original parent of Y is not dirtied. We must detect that there are now logically two Ys in the tree (even though actually there's only one, because it was reparented).

- A widget with key Z is in the tree. Its parent A and an unrelated widget B are both marked dirty. A rebuilds and removes its child, B rebuilds and adds a child with key Z of the same type. We must not report any duplicates.

- A widget with key Z is in the tree. Its parent A and an unrelated widget B are both marked dirty. B rebuilds and adds a child with key Z of the same type, then A rebuilds and removes its child. We must not report any duplicates.

- A widget with key Z is in the tree. Its parent A and an unrelated widget B are both marked dirty. B is a layout builder. A rebuilds and removes its child, B, in a different build scope in the same frame, rebuilds and adds a child with key Z of the same type. We must not report any duplicates.

- A widget with key Z is in the tree. Its parent A and an unrelated widget B are both marked dirty. A is a layout builder. B rebuilds and adds a child with key Z of the same type, then A, in a different build scope in the same frame, rebuilds and removes its child. We must not report any duplicates.

- A widget with multiple children adds two children of the same type and with the same key to the tree. We must report them as duplicates.

- A widget with multiple children including one child with key C reorders its children causing C to be in a different place in the list. We must not report any duplicates.

- A widget with multiple children including one child with key C reorders its children and adds a second child with the same type as key as the one with key C, while simultaneously moving C to a different place in the list. We must report them as duplicates.

- The example listed earlier in this bug. We must not list duplicates.

- More examples.

So long as we have tests for all of these, and so long as our logic is relatively efficient, I don't have a strong attachment to any particular mechanism for implementing the GlobalKey duplicate detection.

**Comment by krisgiesing** [comment]
I wonder if this would be sufficient:
1. track reservations as a list of parents
2. add and remove parents as necessary during the build
3. in `finalizeTree`, assert that the parent-list length is 1 or 0

**Comment by chunhtai** [comment]
@krisgiesing yes I think that will be the approach we want to take. However, the hard part is how do we detect whether we should remove the parent or not.

For example: if I saw a key that has register to a parent before, how do i know if this is indeed a key duplication or this is a result of earlier rebuild.
or when you rebuild the parent, how to detect which key under it is a result of earlier rebuild

I am still currently experimenting different ways to implement it.

**Comment by krisgiesing** [comment]
@chunhtai I'm not sure I quite follow your response. My suggestion is to add and remove parents during the course of the build the same way we do today, not because there is any in-flight duplicate detection, but because the object graph is being processed by client code in the way that the client dictates. Only at the end, once the client has finished constructing the graph to its specification, do we check this consistency rule that keys must have one and only one parent. As I understand it, the main difference between that approach and today's approach is that keys must have a list of parents during the build, whereas today there's only a single parent in the reservation for a key.

**Comment by chunhtai** [comment]
@krisgiesing I might missing something, but this is the issue I am facing.
Consider above case
![image](https://user-images.githubusercontent.com/47866232/70265100-900ed300-174e-11ea-9c52-a0485a68072a.png)
child1 is marked dirty and we switch child1 and child2

This is the sequence we received in _debugReserveFor
1. reserve child 1 for key1 (early rebuild)
2. reserver child 2 for key1 (as you said we should temporary allow this and add child 2 into a list)
3. reserve child1 for key2 (this is where it gets tricky. In this case, we should remove child1 from the list of key1. However, this can be a real duplicate keys issue. If child1 has multiple children, how do we detect which key reservations are the results of early rebuild and which are not)

I can get around this problem by register both child and parent to a global key, and correctly remove the reservation. However, that cause another issue... To reserve a global key with the child, it means i have to build the child first before i reserve. That mean we will have a broken tree if we really have duplicate key issue. Previously, this will be prevented because we throw error before we build the child. This is problematic for test

**Comment by krisgiesing** [comment]
I think I follow you now. The step 2 that I listed, "add and remove children as necessary during the build", assumes that the framework already knows how to classify child movement operations during the build pass, and currently it doesn't.

**Comment by krisgiesing** [comment]
I wonder if the problem is that we are doing the reservation when the child builds, and we should instead be doing the reservation when the parent builds (and clearing the reservation at the start of that process)? Then in your step 3 we would release the reservation of Child1 by key1 when Child1 was building.

**Comment by chunhtai** [comment]
Here are the list of test in the code, note most of them have multiple test except the one i added 

> * A widget with key X is in the tree. Later, an unrelated widget is marked dirty, rebuilds, and adds a new widget with key X (not the same type as the original X). Even though only one widget was added that frame, we must detect that there are two Xs in the tree (without a global tree walk which is prohibitively expensive).
'GlobalKey duplication 19 - subtree build duplicate key with same different type'
> * A widget with key Y is in the tree. An unrelated widget is marked dirty, and it adds a child with key Y, having the same type as the original widget with key Y, causing the original Y's element to be reparented. The original parent of Y is not dirtied. We must detect that there are now logically two Ys in the tree (even though actually there's only one, because it was reparented).
'GlobalKey duplication 18 - subtree build duplicate key with same type'
> * A widget with key Z is in the tree. Its parent A and an unrelated widget B are both marked dirty. A rebuilds and removes its child, B rebuilds and adds a child with key Z of the same type. We must not report any duplicates.
https://github.com/flutter/flutter/blob/0aed0b61a1789ed06c9fa05a6c5fcfbd398c4811/packages/flutter/test/widgets/global_keys_moving_test.dart#L57
> * A widget with key Z is in the tree. Its parent A and an unrelated widget B are both marked dirty. B rebuilds and adds a child with key Z of the same type, then A rebuilds and removes its child. We must not report any duplicates.
https://github.com/flutter/flutter/blob/0aed0b61a1789ed06c9fa05a6c5fcfbd398c4811/packages/flutter/test/widgets/global_keys_moving_test.dart#L57
> * A widget with key Z is in the tree. Its parent A and an unrelated widget B are both marked dirty. B is a layout builder. A rebuilds and removes its child, B, in a different build scope in the same frame, rebuilds and adds a child with key Z of the same type. We must not report any duplicates.
'GlobalKey correct case 2 - can move global key from layoutbuilder to container widget'

> * A widget with key Z is in the tree. Its parent A and an unrelated widget B are both marked dirty. A is a layout builder. B rebuilds and adds a child with key Z of the same type, then A, in a different build scope in the same frame, rebuilds and removes its child. We must not report any duplicates.
'GlobalKey correct case 1 - can move global key from container widget to layoutbuilder'
> * A widget with multiple children adds two children of the same type and with the same key to the tree. We must report them as duplicates.
'GlobalKey duplication 6 - splitting and not changing type'

> * A widget with multiple children including one child with key C reorders its children causing C to be in a different place in the list. We must not report any duplicates.
https://github.com/flutter/flutter/blob/0aed0b61a1789ed06c9fa05a6c5fcfbd398c4811/packages/flutter/test/widgets/global_keys_moving_test.dart#L57
> * A widget with multiple children including one child with key C reorders its children and adds a second child with the same type as key as the one with key C, while simultaneously moving C to a different place in the list. We must report them as duplicates.
'GlobalKey duplication 6 - splitting and not changing type'

**Comment by chunhtai** [comment]
design doc flutter.dev/go/globalkey-duplication-refactoring

**Comment by chunhtai** [comment]
pr reverted due to bad merge
